### DISCUSSION TITLE
Magnetic Calculation(AFM)

### DISCUSSION THREAD

**Discussion by 9yulo** [TRIQS/solid_dmft#17] [Q&A]
Dear TRIQS community,
I want to do an AFM calculation similar to the example you did on [NdNiO2](https://triqs.github.io/solid_dmft/tutorials/NNO_os_plo_mag/tutorial.html).

1-) Doing the NdNiO2 example this took me more than 24h running on 96 nodes. at the level of the analysis of the spectral function, to have the same tracing as here I have to do the maxent analysis?[NiO](https://triqs.github.io/dft_tools/latest/tutorials/nio_csc.html#spectral-function-on-real-axis-maxent) or simply how to have this quantity?

2-) can we do the same AFM calculation with DFTTOOLS? because I launched the calculation in super slow I launched the calculation with solid_dmft (mpirun -np 240 solid_dmft confi.ini ) it has been running for more than 30 hours with a single iteration.

**Comment by the-hampel** [comment]
Hi @9yulo,

I will start answering your second question first. DFTTools is just a library with functions and classes. You have to write your own DMFT loop etc to actually use it. solid_dmft is exactly doing that, using DFTTools to write a DMFT loop. There should be very little overhead in your calculation caused by solid_dmft. Most of the time the code is spending in the solver call (cthyb). In the example you are mentioning roughly 85% of the time are spent there, making this the computational expensive part. However, there is nothing in solid_dmft slowing down your calculations compared to a more minimal setup you might achieve by tailoring the script to your problem. This should clarify a bit your second question.

Regarding your first question: the NdNiO2 tutorial example you are mentioning should take roughly 4-5 hours on 1024 cores. I am not sure how many cores you in the end used? You are mentioning 96 nodes? How many cores does each node have? 

How computationally demanding your task is depends a lot on the system at hand. In general a larger U value will make things easier for cthyb and you can get better results with smaller number of `ncycles`. Or you can maybe reduce the number of orbitals in the active space, or use a less complicated local Hamiltonian structure. The example in the tutorial is rather "hard". For certain edge cases approximate solvers might also help. See for example tutorial `04-HubbardI-comparison.ipynb` on https://github.com/TRIQS/tutorials/tree/3.1.x/AbinitioDMFT . However, solving the impurity problem within the DMFT loop numerically exact, i.e. with cthyb, is a much more demanding that typical DFT calculations. This should be the same for other DMFT implementations out there. 

If you are interested in approximate solvers, we will integrate the latest https://github.com/TRIQS/hartree_fock/ solver in solid_dmft, which will be very fast solving DMFT on the Hartree level, which is the same approximation done in DFT+U. 


Best,
Alex

**Comment by 9yulo** [comment]
Hi @the-hampel,
Thank you for your enlightenment, 

Yes indeed it was 96 cores and not nodes. but the calculation is well finished. and I still have some questions:

1-) for the parameter dc_type = 0 dc = True dc_dmft = False which is Type of double counting correction considered in DFTTOOLS we give the value of the occupation with DC_value. but with this method of solid_DMFT how to be sure that the electrons have converged?  I have by this at the end of the calculation for the example of NdNiO2
total occupany of impurity 0:  8.5264
total occupany of impurity 1:  8.5264
 what shows well an occupation of 8.5 for each impurity at the end of the DMFT calculation what is appreciably equal to the value obtained in the DFT calculation (OUTCAR)
total chargeù

Since in the example we are talking about the AFM calculation, I expected that at the end of the DMFT calculation the two SPINs would be opposite:

how to have the spectral function like here? [NiO](https://triqs.github.io/dft_tools/latest/tutorials/nio_csc.html#spectral-function-on-real-axis-maxent)

**Comment by the-hampel** [comment]
The convergence can be best tested when looking at the `observables_imp0.dat` and `conv_imp0.dat` files. There you should check if convergence is reached over iterations.

The impurity occupations itself in DMFT will greatly depend on the chosen projection method. The orbital occupations in Vasp might be misleading. Those are straight projections for all KS states on atomic d orbitals in the PAW spheres. When using the dfttools vasp converter it will take the same projections, but depending on your chosen energy window only in specific window. This will loose some projection weight, i.e. occupation. More importantly the converter will orthonormalize the projectors, which Vasp is not doing. This will increase the number of electrons again. For the NdNiO2 example you should see around 8.0 occupation per Ni side in the d orbitals, whereas in solid_dmft you have 8.5 electrons. This is reasonably close. 

I urge you to maybe also do the 4 tutorials here: https://github.com/TRIQS/tutorials/tree/unstable/AbinitioDMFT , which use solid_dmft in a very simple example, which you can run even on a normal laptop. It will also show you how to use MaxEnt to calculate spectral functions. For k-dependent spectral functions you have to use at the moment the interface through wannier90. This has technical reasons. However, the calculation of the impurity spectral function is explained in the tutorial I referred to.

Regarding the magnetic moment: solid_dmft only prints the occupation per spin channel. You have to calculate the magnetic moment as difference between n_up - n_down by yourself. In your case: 
```
imp 0 spin up: 4.7519
imp 0 spin down: 3.7745
```
this would lead to a moment of ~1 mu_B. And you can see that with this definition imp0 has +1 mu_B, whereas imp1 has a moment of -1 mu_B. 

Best,
Alex

**Reply by 70akaline** [reply]
Hi, Alex I'm a bit wondering about the magnetisation direction for that antiferromagnetism. Since solid_dmft's implementation is just reverse the spin channel of GFs for different atoms, where the magnetisation point to? Do I need extra rotation to point the magnetisation direction?

**Reply by the-hampel** [reply]
Hi, 
you are absolutely right with that. The magnetization axis is chosen by the shape of the orbitals. Deviations from that choice will show itself in off-diagonal terms in the self-energy (if properly allowed). To my knowledge non-collinear magnetism has not yet really been investigated with DMFT. But basically this is exactly the same way this works in DFT, where the occupation of the up minus down bands gives you the magnetization. off-diagonal terms in the density matrix will then allow for non-collinear magnetism.

**Reply by 70akaline** [reply]
Dear Alex 

Thanks for your answers. I'm currently calculating something like iridate, (Sr2IrO4). The example given by the author of https://github.com/ru-ccmt/EDMFTF uses Wigner transformation to rotate the complex (if my understanding is not wrong) spherical harmonics. I'm just thinking why the Wigner transformation is used, rather than point the antiferromagnetism of Ir atoms through Wien2K

Best regrads.

**Reply by the-hampel** [reply]
Hi, 
I think the way you describe it is done in the tutorial makes totally sense to me. I assume (I never did SOC DMFT calculations myself) that the projector functions used in to downfold from DFT to DMFT are done with a given local or global coordinate frame. This is how the projectors are typically generated, which may vary in a code like Wien2k? Hence, I think that even if you would perform the DFT calculation with magnetism, you would still need to convince the DFT code to rotate the local projections performed to create the projector functions to the coordinate frame via a Wigner transformation. So I think in the end the transformation has to be done at some point anyway. But maybe someone else as a better understanding how this is implemented in Wien2k - this is more my viewpoint.

**Reply by 9yulo** [reply]
Hi @the-hampel,

I finished tutorial 4(AFM) now I have some difficulty to plot the spectral function after doing the maxent. 
i used the maxent_gf_imp.py  [maxent_gf_imp.py.zip](https://github.com/TRIQS/solid_dmft/files/9357073/maxent_gf_imp.py.zip)   and this is finished perfectly.  


```
$ h5ls nno.h5/DMFT_results/last_iter/Aimp_maxent_1/Aimp_w_line_fit
down_0                   Dataset {1, 1, 200, 2}
down_1                   Dataset {1, 1, 200, 2}
down_2                   Dataset {1, 1, 200, 2}
down_3                   Dataset {1, 1, 200, 2}
down_4                   Dataset {1, 1, 200, 2}
up_0                     Dataset {1, 1, 200, 2}
up_1                     Dataset {1, 1, 200, 2}
up_2                     Dataset {1, 1, 200, 2}
up_3                     Dataset {1, 1, 200, 2}
up_4                     Dataset {1, 1, 200, 2}

```
please what value does the 4th index represent?     

I thought it was respectively: dxy dyz dz2 dxz dx2-y2

**Reply by the-hampel** [reply]
Hi, 
The Aimp_w_line_fit is stored as normal numpy array per block. When storing in an h5 archive it is actually stored as complex array so the `2` in the 4th index is for real,complex value. Each block is 1x1x200, where 200 are the number of frequencies used in Maxent. So for example when loading the result via python:
```
with HDFArchive('b10-U3.6/lco_wannier.h5', 'r') as h5:
    Aimp_w = h5['DMFT_results']['last_iter']['Aimp_maxent_0']['Aimp_w_line_fit']
    
for key, val in Aimp_w.items():
    print(key, val.shape)
 ```
 you should get for a paramagnetic case with 1 orbital:
```
total_0 (1, 1, 200)
```
Best,
Alex

**Reply by 9yulo** [reply]
Hi,

have you tested solid DMFT with the latest version of vasp 6.3?
Best,

9yulo

**Reply by the-hampel** [reply]
Hi, yes we did. Everything should work when you apply the patches we provide here: https://github.com/TRIQS/solid_dmft/tree/unstable/Docker . In the folder `vasp_diffs` we provide patch files. You can apply them with the lines shown in the `openmpi_dockerfile`: https://github.com/TRIQS/solid_dmft/blob/fd4167868084ed380eb6ab9d070f2cb8752fa5d6/Docker/openmpi_dockerfile#L122-L126

Solid_dmft will also work without these patches in projector mode. The changes are for CSC calculations, the CSC wannier90 interface, and some performance optimization. The version of the `unstable` branch work with Vasp 6.3.2 and the version on the `3.1.x` branch of solid_dmft work with Vasp 6.3.0.

Best,
Alex

**Reply by 9yulo** [reply]
Hi,
the patch and the compilation is well finished with but the LOCPROJ file is different than what I calculated with vasp.5.4.4.


VASP.5.4.4: LOCPROJ
```
orbital     1     1     1      -52.1286225553        1.0000000000
     1        0.0000002651       -0.0000000221
     2        0.0007498565       -0.0000623912
     3        0.0000039955       -0.0000003324
     4        0.0000430995       -0.0000035861
     5        0.0000004803       -0.0000000400
     6        0.0000001217       -0.0000000101
     7       -0.0001027510        0.0000085493
     8       -0.0000016500        0.0000001373
     9       -0.0003238348        0.0000269444
    10        0.0000000776       -0.0000000065
```
VASP.6.3.0: LOCPROJ
```

orbital     1     1     1      -52.1286225554        1.0000000000
     1       -0.0000002660
     2       -0.0007524477
     3       -0.0000040093
     4       -0.0000432485
     5       -0.0000004820
     6       -0.0000001221
     7        0.0001031061
     8        0.0000016557
     9        0.0003249537
    10       -0.0000000778
```
 **plovasp plo.cfg** 
```
Warning: could not identify MPI environment!
Starting serial run at: 2022-08-24 11:19:50.679546
Traceback (most recent call last):
  File "/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/converter.py", line 89, in <module>
    main()
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/converter.py", line 86, in main
    generate_and_output_as_text(filename, vasp_dir)
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/converter.py", line 57, in generate_and_output_as_text
    vasp_data = vaspio.VaspData(vasp_dir, efermi_required=efermi_required)
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/vaspio.py", line 78, in __init__
    self.plocar.from_file(vasp_dir)
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/vaspio.py", line 141, in from_file
    self.proj_params, self.plo = self.locproj_parser(locproj_filename=vasp_dir + "LOCPROJ")
  File "/RIQS/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/vaspio.py", line 301, in locproj_parser
    ctmp = complex(float(sline[1]), float(sline[2]))
IndexError: list index out of range

```

**Reply by the-hampel** [reply]
Hi,
somehow your LOCPROJ file is missing the complex part (2nd column) and I do not know why. I am not sure how your input looks like, but I just recompiled Vasp 6.3.2 with the patch files from the unstable solid_dmft branch, and rerun this example for SrVO3: https://github.com/TRIQS/solid_dmft/tree/3.1.x/doc/examples/svo-one-shot/converter and I obtain the following correct LOCPROJ: 
```
     1   343    20     5   5.2197220  # of spin, # of k-points, # of bands, # of proj, Efermi
   ISITE:     2    R=     -0.5000000    -0.5000000    -0.5000000  PAW projector    :    dxy   
   ISITE:     2    R=     -0.5000000    -0.5000000    -0.5000000  PAW projector    :    dyz   
   ISITE:     2    R=     -0.5000000    -0.5000000    -0.5000000  PAW projector    :    dz2   
   ISITE:     2    R=     -0.5000000    -0.5000000    -0.5000000  PAW projector    :    dxz   
   ISITE:     2    R=     -0.5000000    -0.5000000    -0.5000000  PAW projector    :   dx2-y2 

orbital     1     1     1      -29.1215272684        1.0000000000
     1        0.0000002807       -0.0000000874
     2       -0.0000003833        0.0000003910
     3       -0.0000008653       -0.0000009103
     4       -0.0000001389        0.0000000433
     5        0.0000002804        0.0000006869
...
```
which looks the same for Vasp 5.4.4. Can you please try this example.  If you don't get this result we can dig deeper.

Best,
Alex

**Reply by 9yulo** [reply]
Hi,

with vasp.6.3.0, its work only with std version not gamma version of vasp.

Best

**Reply by the-hampel** [reply]
Honestly, I never tested the projectors with the gamma point only version of Vasp. Please ask in the Vasp developer forum why this produces different output. I guess this is because one can remove the phase of the wave function if one only has one k-point? But the output should not change in this case, hence this is a bug on the Vasp site.

**Comment by 9yulo** [comment]
Hi,

- There an option to RESTART the calculation with Solid DMFT? if the calculation has stopped let say at the 13th iteration out of 15 for lack of time it is possible to continue the calculation but from the previous calculation?

- After finishing the example of the tutorial 4 what options should be used to use this on-shot calculation to make a self-consistent calculation. 

- How to determine these two parameters? fit_min_w ; fit_max_w ( the max and the min matsubara frequency). 


- I am curious to make a comparison between the Anti-ferromagnetic and ferromagnetic system of the example 4 which parameters to use? magnetic = True ; magmom = +0.01, +0.01 afm_order = False ??

**Reply by the-hampel** [reply]
Hi, 
* to restart a calculation, just re-submit the same job, with the job output folder present. If the name of the jobfolder already exists in the current dir, solid_dmft will try to continue the calculation and add another `n_iter_dmft` cycles to it. 
* Not sure what you mean with the second point? you have to enable the csc=True flag. Try to follow this example for the additional flag: https://triqs.github.io/solid_dmft/tutorials/Ce2O3_csc_w90/tutorial.html
* you have to do the calculation first with some educated guess for the tail fitting parameters. Then plot the resulting self-energy and manually use the tail fitting routines in triqs (an example can be found here: https://github.com/the-hampel/pystack/blob/a0b03d9db1e945ee0f94f07130a597d43e4152d1/plot_triqs.py#L37) to test different parameters to give a smooth self-energy for the system at hand. 
* Yes, just can either put afm_order=False. An advanced option is to use afm_order=True and put the same magmom on the impurities, i.e. 
```
magnetic = True 
magmom = +0.01, +0.01 
afm_order = True
```
This will enforce a FM solution on the two impurities. But it is always advisable to cross-check with a afm_order=False calculation if this is a stable solution (no symmetries used)

Best,
Alex
